### ISSUE TITLE
How to get average of absolute gradient during the training

### ISSUE THREAD

**Issue by junwen-austin** [PyTorchLightning/pytorch-lightning#5238]
## ❓ Questions and Help
I would like to calculate the average of absolute value of gradients for a neural network during the training so that I make sure gradients do not vanish or explode during the training. What would be the way to do it? 

In Pytorch,  one can do something like below to collect the gradient after `backward()` is called:

```
optimizer.zero_grad()

pred = model(input)
loss = loss_fn(pred, y)

loss.backward()
grads = []
for param in model.parameters():
    grads.append(param.grad.view(-1))

grads = torch.cat(grads).abs().mean()

```

Since PyTorch Lightning takes care of `backward` behind the scene, I really appreciate it if someone could help. Thanks.

**Comment by awaelchli** [comment]
There are probably several ways. One is by overriding on_after_backard hook on the LighntingModule. I assume you just want to log the value?

**Comment by junwen-austin** [comment]
@awaelchli  Thank you so much! Yes I just wanted to log the average of absolute gradient for each parameter. Looking at the example from PyTorch Lightning:
https://pytorch-lightning.readthedocs.io/en/stable/lightning_module.html#on-after-backward

```
def on_after_backward(self):
    # example to inspect gradient information in tensorboard
    if self.trainer.global_step % 25 == 0:  # don't make the tf file huge
        params = self.state_dict()
        for k, v in params.items():
            grads = v
            name = k
            self.logger.experiment.add_histogram(tag=name, values=grads,
                                                 global_step=self.trainer.global_step)
```

Should we have `grads = v.grad` 

instead of `grad = v` in the example above? 

My understanding is v itself gives the value of the parameter while v.grad gives the gradient.

Thanks again!

**Comment by rohitgr7** [comment]
yep should be `v.grad`. mind send a PR?

**Comment by Tikquuss** [comment]
> @awaelchli Thank you so much! Yes I just wanted to log the average of absolute gradient for each parameter. Looking at the example from PyTorch Lightning: https://pytorch-lightning.readthedocs.io/en/stable/lightning_module.html#on-after-backward
> 
> ```
> def on_after_backward(self):
>     # example to inspect gradient information in tensorboard
>     if self.trainer.global_step % 25 == 0:  # don't make the tf file huge
>         params = self.state_dict()
>         for k, v in params.items():
>             grads = v
>             name = k
>             self.logger.experiment.add_histogram(tag=name, values=grads,
>                                                  global_step=self.trainer.global_step)
> ```
> 
> Should we have `grads = v.grad`
> 
> instead of `grad = v` in the example above?
> 
> My understanding is v itself gives the value of the parameter while v.grad gives the gradient.
> 
> Thanks again!

I tried to get the gradient as you do, and `v.grad` was always equal to `None`.

Here's how I collected the gradient.

```python
def on_after_backward(self):
    if self.trainer.global_step % 1 == 0:  
        grad_vec = None
        for p in self.parameters():
            #p.grad.data.div_(batch["input"].shape[0])
            if grad_vec is None:
                grad_vec = p.grad.data.view(-1)
            else:
                grad_vec = torch.cat((grad_vec, p.grad.data.view(-1)))
                
        print(grad_vec.shape)
```

`pytorch_lightning==1.6.3`
